src/main/scala/nl/biopet/tools/tenxkit/groupdistance/GroupDistance.scala

```diff
@@ -73,27 +74,36 @@ object GroupDistance extends ToolCommand[Args] {
 //          correctCellsMap,
 //          50000000)
 //    }
-    val vectors = distanceMatrixToVectors(distanceMatrix.value, correctCells).toDF("sample", "features").cache()
 
-    val bkm = new BisectingKMeans()
-      .setK(cmdArgs.numClusters)
-      //.setMaxIter(cmdArgs.numIterations)
-      .setSeed(cmdArgs.seed)
-
-    val model = bkm.fit(vectors)
-
-    val predictions: RDD[(Int, Iterable[Int])] = model
-      .transform(vectors)
-      .select("sample", "prediction")
-      .as[Prediction]
-      .rdd
-      .groupBy(_.prediction)
-      .repartition(cmdArgs.numClusters)
-      .map { case (group, list) => group -> list.map(_.sample) }
-      .cache()
+    val initGroups: RDD[GroupSample] = if (cmdArgs.skipKmeans) {
+      val random = new Random(cmdArgs.seed)
+      sc.parallelize(correctCells.value.indices.map(i => GroupSample(1, i)),
+                     correctCells.value.length)
+    } else {
+      val vectors = distanceMatrixToVectors(distanceMatrix.value, correctCells)
+        .toDF("sample", "features")
+        .cache()
+
+      val bkm = new BisectingKMeans()
+        .setK(cmdArgs.numClusters)
+        //.setMaxIter(cmdArgs.numIterations)
+        .setSeed(cmdArgs.seed)
+
+      val model = bkm.fit(vectors)
+
+      model
+        .transform(vectors)
```

Comment:
Could you use some intermediate variables here? By `.as[Prediction]` I am completely lost on what you are doing.